src/main/java/fr/sncf/osrd/utils/DoubleRangeMap.java

```diff
@@ -26,7 +26,7 @@ public void addRange(double begin, double end, double value) {
             lastValue = floorEntry.getValue();
 
         for (var entry : subMap(begin, end).entrySet()) {
-            if (entry.getKey() >= begin)
+            if (entry.getKey() <= begin)
```

Comment:
This is what was causing the bug, wrongly evaluating the slope under the train. That sign was in the wrong direction so only the floor entry was returned, rather than all the slopes under the train.

Reply:
It could even be changed to `!=` now, since we can't have any value lower than `begin` here.